llvm/include/llvm/CAS/Namespace.h

```diff
@@ -0,0 +1,87 @@
+//===- llvm/CAS/Namespace.h -------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CAS_NAMESPACE_H
+#define LLVM_CAS_NAMESPACE_H
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Error.h"
+
+namespace llvm {
+namespace cas {
+
+class UniqueID;
+class UniqueIDRef;
+
+/// Namespace for the CAS. Has an identifier and knows its hash size.
+///
+/// Instances should be declared \a ManagedStatic to compare by pointer
+/// identity.
+class Namespace {
```

Comment:
I wonder why we need Namespace instead of just put everything in `CASDB` and reference CASDB instead?

Reply:
Two CAS instances can be in the same namespace. For example, consider creating the following five builtin CAS instances:
```
std::unique_ptr<ObjectStore> CASes[] = {
    createInMemoryCAS(),
    createInMemoryCAS(), // independent in-memory CAS instance
    createOnDiskCAS("/path/1"), // on-disk storage
    createOnDiskCAS("/path/1"), // same on-disk storage,  but different in-memory instance
    createOnDiskCAS("/path/2"), // different on-disk storage
};
```
These assertions should succeed:
```
const Namespace &NS = CASes[0].getNamespace(); 
for (auto &CAS : llvm::make_range(CASes))
  assert(&NS == &CAS.getNamespace());
```

This property seems critical to me for decoupling `ActionCache` from `ObjectStore`. Otherwise, how does `ActionCache` return a `UniqueID`, if the `UniqueID` needs to reference an `ObjectStore`?